doc/contributing/adding-v8-fast-api.md

```diff
@@ -0,0 +1,153 @@
+
+## Requirements
+
+* Each unique fast API function signature should be defined inside the
+  [`external references`](../../src/node_external_reference.h) file.
+* To test fast APIs, make sure to run the tests in a loop with a decent
+  iterations count to trigger V8 for optimization and to prefer the fast API
+  over slow one.
+
+## Fallback to slow path
+
+Fast API supports fallback to slow path in case logically it is wise to do so,
+for example when providing a detailed error, or need to trigger JavaScript.
+Fallback mechanism can be enabled and changed from both the caller
+JavaScript function or from the fast API function declaration.
+
+Every fast API function accessible from JavaScript side can pass an object
+consisting of fallback key with a boolean value as the last parameter
+(or the first parameter, if no parameters of the function exist).
+
+Note that, the fallback options parameter does not need to be defined
+in the C++ function in order for it to be used in the JavaScript side.
+
+In V8, the options fallback is defined as `FastApiCallbackOptions` inside
+[`v8-fast-api-calls.h`](../../deps/v8/include/v8-fast-api-calls.h).
+
+* JavaScript land
+
+  Example of a JavaScript call into a fast API:
+
+  ```js
+  // Let divide be a function that provides fast api calls.
+  const { divide } = internalBinding('custom_namespace');
+
+  function divide(a, b) {
+    return divide(a, b, { fallback: a === 0 })
+  }
+  ```
+
+* C++ land
+
+  Example of a conditional fast path on C++
+
+  ```cpp
+  // Anywhere in the execution flow, you can set fallback and stop the execution.
+  static double divide(const int32_t a,
+                       const int32_t b,
+                       v8::FastApiCallbackOptions& options) {
+    if (b == 0) {
+      options.fallback = true;
+    } else {
+      return a / b;
+    }
+  }
+  ```
+
+## Example
+
+A typical function that communicates between JavaScript and C++ is as follows.
+
+* On the JavaScript side:
+
+  ```js
+  const { divide } = internalBinding('custom_namespace');
+  ```
+
+* On the C++ side:
+
+  ```cpp
+  #include "v8-fast-api-calls.h"
+
+  namespace node {
+  namespace custom_namespace {
+
+  static void divide(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+    CHECK(args[0]->IsInt32());
+    CHECK(args[1]->IsInt32());
+    auto a = args[0].As<v8::Int32>();
+    auto b = args[1].As<v8::Int32>();
+
+    if (b->Value() == 0) {
+      return node::THROW_ERR_INVALID_STATE(env, "Error");
+    }
+
+    double result = a->Value() / b->Value();
+    args.GetReturnValue().Set(result);
+  }
+
+  static double FastDivide(const int32_t a,
+                           const int32_t b,
+                           v8::FastApiCallbackOptions& options) {
+    if (b == 0) {
+      options.fallback = true;
+    } else {
+      return a / b;
+    }
+  }
+
+  CFunction fast_divide_(CFunction::Make(FastDivide));
```

Comment:
According to https://github.com/nodejs/node/blob/66a89777280c1ba431ba467cb241a6b7c0112704/deps/v8/include/v8-fast-api-calls.h#L33-L36, it seems like accepting a `FastApiCallbackOptions` would only be possible if you use `MakeWithFallbackSupport` instead of `Make`. How does this work? FWIW, I tried looking for the definition of `MakeWithFallbackSupport` but I wasn't able to find it.

Reply:
I couldn't find a reference to it either. I think this might be outdated. Better check for it. @devsnek Any idea?

Reply:
Looking at https://source.chromium.org/chromium/_/chromium/v8/v8.git/+/46e349bb83ea96bd63b10c659a72e3821c98c496:test/cctest/test-cpu-profiler.cc;l=4493;drc=539b50f5ae2f793156ecdae679e18fe1214e57cc;bpv=0;bpt=0 and https://source.chromium.org/chromium/_/chromium/v8/v8.git/+/46e349bb83ea96bd63b10c659a72e3821c98c496:test/cctest/test-cpu-profiler.cc;l=4385-4386;drc=539b50f5ae2f793156ecdae679e18fe1214e57cc;bpv=0;bpt=0, it sounds like it is expected to use `Make()`, so I agree that it's probably an outdated comment.